tools/bot/matrix.go

```diff
@@ -0,0 +1,147 @@
+}
+
+// LoginResponse is the response for a login request
+type LoginResponse struct {
+	UserId      string `json:"user_id"`
+	AccessToken string `json:"access_token"`
+	HomeServer  string `json:"home_server"`
+	DeviceId    string `json:"device_id"`
+	WellKnown   struct {
+		MHomeserver struct {
+			BaseUrl string `json:"base_url"`
+		} `json:"m.homeserver"`
+	} `json:"well_known"`
+}
+
+var (
+	// Base Links
+	clientUrl = "https://" + tools.Cfg.Alerts.Matrix.Homeserver + "/_matrix/client/r0/"
+	loginUrl  = clientUrl + "login"
+)
+
+const (
+	// Maximum transaction ID
+	maxID = 10000
+
+	// Suffixes and Prefixes for sending a message to a room
+	accessTokenSuffix = "?access_token="
+	roomSuffix        = "rooms/"
+	roomMsgPrefix     = "/send/m.room.message/"
+
+	// Constants describing the messages itself
+	msgType     = "m.text"
+	msgFormat   = "org.matrix.custom.html"
+	loginMethod = "m.login.password"
+
+	contentType = "application/json"
+)
+
+// SendBotMessage sends a formatted message to a matrix room
+func (ma *Matrix) SendBotMessage(info InfoMessage) error {
+	id := strconv.Itoa(rand.Intn(maxID)) // transaction id
+	authToken, err := getAuthToken()
+	if err != nil {
+		return err
+	}
+	if authToken == "" {
+		return errors.New("authentication failed, could not get token")
+	}
+
+	var roomMessageSuffix = roomMsgPrefix + id + accessTokenSuffix
+	url := clientUrl + roomSuffix + tools.Cfg.Alerts.Matrix.RoomID + roomMessageSuffix + authToken
+	matrixMessage := Message{
+		MsgType:       msgType,
+		Body:          generateInfoText(info),
+		Format:        msgFormat,
+		FormattedBody: getFormattedMessageText(info),
+	}
+	matrixMessageJSON, err := json.Marshal(matrixMessage)
+	if err != nil {
+		return err
+	}
+	err = sendMessageRequest(url, bytes.NewBuffer(matrixMessageJSON))
+	return err
+}
+
+// sendMessageRequest sends a PUT request to url with a given body
+func sendMessageRequest(url string, body io.Reader) error {
+	client := &http.Client{}
+	request, err := http.NewRequest(http.MethodPut, url, body)
+	if err != nil {
+		return err
+	}
+	response, err := client.Do(request)
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf(fmt.Sprintf("received status code %d instead of %d.", response.StatusCode, http.StatusOK))
+	}
+	return err
+}
+
+// getAuthToken retrieves a single use token for the next message sent to the server.
+func getAuthToken() (string, error) {
+	login := Login{
+		Type:     loginMethod,
+		User:     tools.Cfg.Alerts.Matrix.Username,
+		Password: tools.Cfg.Alerts.Matrix.Password,
+	}
+	loginRequest, err := json.Marshal(login)
+	if err != nil {
+		return "", err
+	}
+	response, err := http.Post(loginUrl, contentType, bytes.NewBuffer(loginRequest))
+	if err != nil {
+		return "", err
+	}
+	if response.StatusCode != http.StatusOK {
+		return "", fmt.Errorf(fmt.Sprintf("received status code %d instead of %d.", response.StatusCode, http.StatusOK))
+	}
+	responseBody, err := io.ReadAll(response.Body)
```

Comment:
You can use jsons Decoder to unmarshal the reader https://pkg.go.dev/encoding/json#NewDecoder